analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/LocksReleasedAllPathsBase.cs

```diff
@@ -154,8 +183,33 @@ private ProgramState RemoveLock(SymbolicContext context, ISymbol symbol)
         private static ProgramState AddLock(SymbolicContext context, IOperation operation) =>
             context.State.SetOperationConstraint(operation, context.SymbolicValueCounter, LockConstraint.Held);
 
-        private static ISymbol FirstArgumentSymbol(IInvocationOperationWrapper invocation) =>
-            IArgumentOperationWrapper.FromOperation(invocation.Arguments.First()).Value.TrackedSymbol();
+        private static ISymbol ArgumentSymbol(IInvocationOperationWrapper invocation, int argumentPosition) =>
+            IArgumentOperationWrapper.FromOperation(invocation.Arguments[argumentPosition]).Value.TrackedSymbol();
+
+        private static bool IsInvocationWithBoolRefParam(IInvocationOperationWrapper invocation, KnownType type, int numberOfArguments, int refParameterPosition, params string[] methodNames) =>
+            invocation.TargetMethod.IsAny(type, methodNames)
+            && invocation.Arguments.Length == numberOfArguments
+            && invocation.TargetMethod.Parameters[refParameterPosition].IsType(KnownType.System_Boolean);
+
+        private ProgramState[] HeldAndNotHeldStates(SymbolicContext context, ISymbol lockSymbol, ISymbol refParameter)
+        {
+            ProgramState lockHeldState;
+            ProgramState lockNotHeldState;
+            var refParamSymbolicValue = context.State[refParameter];
+            lockHeldState = AddLock(context, lockSymbol);
+            if (refParamSymbolicValue == null)
```

Comment:
I don't understand the `if`. the `SetSymbolConstraint` should do the whole trick for us on one call. Both, `if` and `else` branches seems to do the same thing